torch/csrc/autograd/record_function.cpp

```diff
@@ -26,39 +25,57 @@ void popCallback() {
   }
   start_callbacks.pop_back();
   end_callbacks.pop_back();
-  has_callbacks = !start_callbacks.empty();
+  if (callback_needs_inputs > 0) {
+    --callback_needs_inputs;
+  }
+}
+
+bool hasCallbacks() {
+  return !start_callbacks.empty();
 }
 
-RecordFunction::RecordFunction(Function* fn, GetPackedInputsCallback cb) {
-  if (!has_callbacks) {
+RecordFunction::RecordFunction() : needs_inputs_(callback_needs_inputs > 0) {}
```

Comment:
Why do we stash `needs_inputs_`? Do we really expect that someone will keep popping callbacks in between the construction and destruction of `RecordFunction`?